conan/tools/gnu/pkgconfigdeps.py

```diff
@@ -0,0 +1,203 @@
+Version: 1.2.3
+Libs: -L${libdir} -lmy-project-1 -linkerflag -Wl,-rpath=${libdir}
+Cflags: -I${includedir}/my-project-1
+Requires: glib-2.0 >= 2.40 gio-2.0 >= 2.42 nice >= 0.1.6
+Requires.private: gthread-2.0 >= 2.40
+"""
+
+
+def _concat_if_not_empty(groups):
+    return " ".join([param for group in groups for param in group if param and param.strip()])
+
+
+def get_target_namespace(req):
+    ret = req.new_cpp_info.get_property("pkg_config_name", "PkgConfigDeps")
+    return ret or req.ref.name
+
+
+def get_component_alias(req, comp_name):
+    if comp_name not in req.new_cpp_info.components:
+        # foo::foo might be referencing the root cppinfo
+        if req.ref.name == comp_name:
+            return get_target_namespace(req)
+        raise ConanException("Component '{name}::{cname}' not found in '{name}' "
+                             "package requirement".format(name=req.ref.name, cname=comp_name))
+    ret = req.new_cpp_info.components[comp_name].get_property("pkg_config_name", "PkgConfigDeps")
+    return ret or comp_name
+
+
+class PkgConfigDeps(object):
+
+    def __init__(self, conanfile):
+        self._conanfile = conanfile
+
+    @property
+    def compiler(self):
+        return self._conanfile.settings.get_safe("compiler")
+
+    def _get_require_comp_name(self, pkg_name, req):
+        pkg, comp_name = req.split("::") if "::" in req else (pkg_name, req)
+        req = self._conanfile.dependencies.direct_host[pkg]
+        cmp_name = get_component_alias(req, comp_name)
+        return cmp_name
+
+    def get_components(self, pkg_name, dep):
+        ret = []
+        for comp_name, comp in dep.new_cpp_info.get_sorted_components().items():
+            comp_genname = get_component_alias(dep, comp_name)
+            comp_requires_gennames = []
+            for require in comp.requires:
+                comp_requires_gennames.append(self._get_require_comp_name(pkg_name, require))
+            ret.append((comp_genname, comp, comp_requires_gennames))
+        return ret
+
+    def _get_public_require_deps(self, comp):
+        public_comp_deps = []
+        for require in comp.requires:
+            if "::" in require:  # Points to a component of a different package
+                pkg, cmp_name = require.split("::")
+                req = self._conanfile.dependencies.direct_host[pkg]
+                public_comp_deps.append(
+                    (get_target_namespace(req), get_component_alias(req, cmp_name)))
+            else:  # Points to a component of same package
+                public_comp_deps.append((get_target_namespace(self._conanfile),
+                                         get_component_alias(self._conanfile, require)))
+        return public_comp_deps
+
+    @property
+    def content(self):
+        ret = {}
+        host_req = self._conanfile.dependencies.host
+        for require, dep in host_req.items():
+            pkg_genname = get_target_namespace(dep)
+
```

Comment:
@memsharded 
I removed the old `self._validate_components(cpp_info)` because I think it's not needed here but let me know your opinion

Reply:
Yes, don't worry about components validation